### ISSUE TITLE
Does the package work for non-exclusive multi-class classification ? 

### ISSUE THREAD

**Issue by NakedKoala** [mahmoodlab/CLAM#47]
looking at `tumor_subtyping_dummy_clean.csv`  The subtyping problem seems to have label class being exclusive.

I am trying to apply this package to a problem where a slide can simultaneously belong to multiple classes.

### Question1:

Is there any particular factors from the architecture side that would prevent my non-exclusive multi-class classification setup from achieving the good results in the exclusive multi-class classification setup (ie. like the multi-class subtyping stuff investigated in the paper)  ?

### Question2:

In non-exclusive multi-class classification setup , my label is a vector of (1, num_class) containing individual class probabiliyt.
I will need to replace the `CrossEntropyLoss` with `BCELoss`

Do you see any reason that this change would break the entire architecture ?

**Comment by fedshyvana** [comment]
For 1, I would recommend trying it first with --no_inst_cluster flag enabled to avoid any assumptions about there being a single class on each slide. Other than that you should be ok. 

For 2, I'm not familiar with multi-label classification, but I imagine you also need to change softmax ---> sigmoid for inference? Although for optimization maybe the activation is already implicitly built into the pytorch loss functions (just double check that this is the case) so yeah you would just have to change the loss function.

**Comment by NakedKoala** [comment]
Thanks for the quick reply.

Are you suggesting that Instance clustering relies on both instance  and slide belonging to an exclusive class ? 

I understand that if instance has multiple-class, then it might be hard to do clustering.  

But why does instance clustering relying on the slide being single class ? 

I am working on a problem where I have slide level label but not the instance level label. I am trying to leverage slide level classification training and instance clustering  ( assuming each instance has exclusive label ) to perform label discovery on the instance patch. 

Given my objective, is there modification that I can do ?

**Comment by fedshyvana** [comment]
ok if I understand correctly, you can try instance level clustering with both clam_sb (single attention branch) and clam_mb (multiple attention branches, one for each class) and compare.

Instance clustering as its implemented now, uses a binary one-vs-rest classification problem for each class in your problem. So lets say there's 3 classes: a, b, c at the slide-level, it will use one classifier to separate a from non-a (which could include b, c but also other junk like pen-markings, class-agnostic cell types, etc.), b from non-b, c from non-c. This is based on the assumption that if you have a slide with slide-level label == a, for example, then the high attention patches should be indicative/characteristic of class a, and thus return positive for the a-classifier, but negative for the b and c-classifiers. 

However, if you are doing multi-label training, where each slide can take on multiple classes, and you use a single attention branch (clam_sb), the high attention patches might belong to multiple different classes, therefore, it's unclear whether it would make sense to put them in the same cluster. This however, should be ok if you use clam_mb (one branch for each class), since there, you're using one branch per class, and assuming the high attention patches from each branch correspond exclusively to only that class, so you can still iterate over all the branches, and cluster the high/low attention branches from each one separately. I believe there shouldn't be any modification you have to make to run it, but lemme know if you run into issues.
